### PR TITLE
Feature/library autocomplete

### PR DESCRIPTION
**Pull request by juliankrispel**
Hello @remy. Here is the promised pr. Please review.

I hadn't much time to do this so it's likely I didn't follow your conventions. 

A couple of questions:
- Running the jshint task I'm getting a ton of errors, is there something I'm missing? 
- What browsers do you support? So far I hand-tested in late firefox, safari and chrome.
- no unit-tests?
- are you considering using react or something of the likes for jsbin? Just curious, would make it a lot more fun contributing.

jsbin-libraries

Also: Thanks so much for jsbin, `tis awesome!

:heart:

### CODE REVIEW

**Comment by remy** [comment]
I've not checked the branch yet to play - but I will do. Here are initial responses to questions:

> Running the jshint task I'm getting a ton of errors, is there something I'm missing?

Yeah, ignore the whole thing. Grunt is *only* and *exclusively* used to build the min file when going to production. Nothing else. It's just a hangover from legacy code.

> What browsers do you support? So far I hand-tested in late firefox, safari and chrome.

IE8+ and latest-1 on all others. Really the (hand) testing is focused on the DOM, because that's where the quirks are. 

> no unit-tests?

Yup. Welcome to legacy 101. I have a blog post about my tests these days, it's a different situation. If you wanted to take on a PR for tests, you'd pretty much win the Internet!

> are you considering using react or something of the likes for jsbin? Just curious, would make it a lot more fun contributing

Not at all. If I were looking at a framework, it would be more likely polymer due to it's alignment to the Shadow DOM. JS Bin is coming up to 7 years old. I'm prepared to bet money that React code wouldn't work in another 7 years, if it even exists still...

But yeah, I definitely put my hands up (and do so publicly) to say that *most* of the client side code is a mess (the cross file globals is particularly fun, as is the `jsbin.panels.panels...` :) )

**Comment by juliankrispel** [comment]
@remy 7 years? WHOW! I had no idea. That's pretty darn amazing :D

File: public/js/editors/library.js [hunk 1]
```diff
@@ -2,69 +2,111 @@
...
+  // save libraries whose label match the search
+  var filteredLibraries = _.filter(libraries, function(library){
+    return library.label.toLowerCase().indexOf(token) > -1 || 
+      librarySearch.trim().length < 1
+  });
 
-  analytics.library('select', group.libraries[selected[1]].label);
-  insertResources(library.url);
-  if (library.snippet) {
-    insertSnippet(library.snippet);
```
Comment (remy) [inline public/js/editors/library.js]: You've lost snippet support in this line.

File: public/js/editors/library.js [hunk 2]
```diff
@@ -2,69 +2,111 @@
...
+  var optionsCount = filteredLibraries.length-1
+
+  // reset the selected item if we've moved beyond the array
+  if(selectedItem > optionsCount && optionsCount > 1){
+    selectedItem = filteredLibraries.length-1;
+  }else if(optionsCount === 0){
+    selectedItem = 0;
   }
-}).on('click', function () {
-  analytics.library('open');
```
Comment (remy) [inline public/js/editors/library.js]: Need to restore analytics

**Comment by remy** [comment]
Initial testing feedback:

- [ ] If I use the cursor keys beyond the size of the box, the highlighted library isn't visible (i.e. press down cursor a few times until the highlight is missing)
- [ ] Since grouping is now removed, it should probably be ordered alphanumerically (case-insensitive)
- [ ] If the user filters the results using the input, is it worth show the number of matches [ala the help menu](https://cloudup.com/cocgvzWgSAQ)? (I'm unsure if this is the right thing to do)
- [ ] Remove dividing lines between the results
- [ ] Library search box height should match navigation (that it sits in) - it looks like it's shy by 1px.
- [ ] Restore the snippet support (commented on the line that shows it being lost)
- [ ] Hitting esc should put the cursor focus back to the editor (to test: focus the HTML panel editor, click add library, then press esc - focus is lost)
- [ ] Currently breaks processors that rely on a library pre-pop (for example, on jsbin.com change JavaScript to JSX) - it uses the `#library` element here: https://github.com/jsbin/jsbin/blob/master/public/js/processors/processor.js#L204
- [ ] How does this behave on a mobile device?
- [ ] Opening the library dropdown should close other dropdowns (just call `hideOpen()`)
- [ ] Can you remove the build of public/js/embed.min.js from the PR.
- [ ] When the library selection is re-opened, the filter should be cleared (i.e. open, search for "promise", select, re-open library, only promise is visible, but nothing in the input box).

General question:

- Should the dropdown open to a more information panel saying to search for your favourite library, rather than listing right away? The idea there is that the code would look up from the libraries.js file but also make look ups (beyond the scope of this PR) to cdnjs et al.

Otherwise, really solid start. Couple of breakages, but hopefully straight forward to fix or patch up.

**Comment by juliankrispel** [comment]
thanks @remy, I'll get to this soon :)

**Comment by getsetbro** [comment]
In jsbin if you add a library it adds the dependencies for you - like jquery. Cdnjs api does not tell you the dependencies of a lib
Also the base api call gives you all the libraries - 1400 right now. But it just gives you 1 URL for that library. For example - for bootstrap you get the main JS file. But what you wanted was the CSS file. So you'd need to request the rest of the bootstrap files from cdnjs, but that would be too many - because it includes the min versions and full versions. So you should have to show the user the rest of them and have them pick the ones they want.

**Comment by remy** [comment]
Just checking in, how we doing over here?

**Comment by juliankrispel** [comment]
Been swamped with work lately, and child care ;). I'm off to SF for 10 days which means I will get some time to work on stuff like this :+1:

**Comment by juliankrispel** [comment]
... within the next few days probably

**Comment by juliankrispel** [comment]
@remy I see that another PR is related, I might start adressing all this on the plane tomorrow morning ;)

**Comment by remy** [comment]
Ha, cheers. Also, child care > (in an inordinate number of magnitudes) PRs!

Cheers.

**Comment by juliankrispel** [comment]
@remy I've scanned the codemirror api docs a little but I'm struggling to find a way to actually focus back on the editor. I thought I'd better ask if you have an idea of how this could be done? Would appreciate some help with that.

**Comment by remy** [comment]
If you scan jsbin code for something like jsbin.panels.focused.focus() it
might be what you want.

**Comment by juliankrispel** [comment]
What's done:
- [x] the library view now scrolls with the highlight.
- [x] Search results are sorted alphabetically.
- [x] Removed the dividers between options (good call)
- [x] Library search box now alignes with the navigation
- [x] Restored snippet support
- [x] Restore jsx support
- [x] Opening the library dropdown should close dropdowns and vice versa
- [x] Removed the embed.min.js from PR
- [x] reset filter when re-opening library dropdown

What's still to do:
- [ ] show number of results
- [ ] Test on mobile devices
- [ ] Refocus on editor after choosing lib
- [ ] Instead of showing libraries straight away, show a mesage saying "search for your favourite libs" or something like that...

**Comment by juliankrispel** [comment]
@remy I tried that and couldn't get it to work :(

**Comment by juliankrispel** [comment]
Right, I gotta hop on a plain now, see you laters

image

**Comment by abhisekp** [comment]
@juliankrispel any updates? This would be a really nice addition :smirk_cat:

**Comment by juliankrispel** [comment]
Dude, I was sitting behind you the other day. Talked to Danny and was like,
damn I need to finish this. I'll get on this in the coming weeks

**Comment by juliankrispel** [comment]
Apologies if this is a little creepy, I should have introduced myself :D

**Comment by juliankrispel** [comment]
Whoopsey, I mistook you for remy, making this even more awkward

Julian Krispel-Samsel
rainforestqa.com
goodafternoon.co

**Comment by abhisekp** [comment]
Please don't remove grouping.
Grouping helps e.g. check out http://www.jsdelivr.com/?query=lodash

Grouping actually helps for visual clarity in the pile of libraries.

**Comment by michalczaplinski** [comment]
hey, I was just thinking of sending a PR with the same enhancements as you @juliankrispel! :) Did you manage to get it finished?

**Pull request closed without merging**